blog/2020-10-08-debezium-community-stories-with-renato-mefi.adoc

```diff
@@ -0,0 +1,143 @@
+* _Manageable effort_; how can we "easily" get 30+ engineers to adopt a solution for this problem? At the same time, how can you ensure the implementation guarantees among every service and team?
+
+We realized that the https://microservices.io/patterns/data/transactional-outbox.html[outbox pattern] would help us meet those requirements: applications would publish events via an outbox table, which gets written to as part of business transactions in the database.
+
+As with the strangler pattern, we wanted to resort to an architecture component, something the teams could self-service. At first, we were exploring a home-grown solution which would look for the outbox tables among every service and publish the messages. The problem with this approach would be the polling databases problem, although in this case this is less harmful as we don't need to look for updates or deletes.
+
+Luckily, by that time I was closely following the work being done in Debezium and I read the blogpost about link:/blog/2019/02/19/reliable-microservices-data-exchange-with-the-outbox-pattern/[reliable data exchange between microservices using the outbox pattern], and there was my answer! Well, I mean, parts of the answer, we still needed to implement it, and that's a story for the next question!
+
+Fast forward a couple of months and we got a reliable way to exchange messages between services, with all the guarantees we wanted to have, and by applying some platform DeVops flavor to it, we also made it self-service and easy to plug in every service!
+
+*You're not only using Debezium but you've also contributed to the project. How was your experience doing so? Are you doing other open-source work, too?*
+
+As I spoiled at the beginning, my usage and contributions to Debezium walked hand-to-hand, in both the use cases we have for Debezium in SurveyMonkey I had great opportunities to contribute to both Debezium and Kafka (just a bug fix, but I'm happy about it!).
+
+At first I was fixing bugs in the Debezium MongoDB connector; as we really scaled it up to all the teams a lot of edge cases started to show up, mostly in the transform which takes the raw database transaction log and transforms into a nicely readable Kafka Connect struct. Also due to our architecture choice, we split the raw log and transformed data into two different steps, which go in separate topics and are configured as separate Kafka Connect connectors. That also created some extra tiny features and challenges to be incorporated in Debezium itself!
+
+Quick sidestep: the rationale behind this decision was to be able to survive transformation errors; MongoDB has a replication window which, if you lose it, means that you are going to have to make a new full snapshot of the collection and you might lose deletion events in this process. Because of this we opted for a safer approach, which was to split the logic of transformation from the raw logs.
```

Comment:
```suggestion
At first, I was fixing bugs in the Debezium MongoDB connector; as we really scaled it up to all the teams a lot of edge cases started to show up, mostly in the transformation which takes the raw database transaction log and transforms into a nicely readable Kafka Connect struct. Also due to our architecture choice, we split the raw log and transformed data into two different steps, which go in separate topics and are configured as separate Kafka Connect connectors.

Quick sidestep: the rationale behind this decision was to be able to survive transformation errors; MongoDB has a replication window which, if you lose it, means that you are going to have to make a new full snapshot of the collection and you might lose deletion events in this process. Because of this we opted for a safer approach, which was to split the logic of transformation from the raw logs like this:
The step we call `op` (stands for operation), is the Debezium MongoDB sink connector and outputs the raw data into the topic without any change or transformation, minimizing the chances of errors in the process, the second step called `cdc`, is a link:https://github.com/salesforce/mirus[Salesforce Mirus] source connector, which reads from the `op` output topic, transforms the message using the link:https://debezium.io/documentation/reference/1.3/configuration/mongodb-event-flattening.html[Debezium Document SMT] and outputs to the final topic, which the services can consume from. With this approach, we now have two main abilities: Resist to errors and crashes on the native/custom transformation process like mentioned above, and have the chance to change the transformation to our desires without having to read from the database again, given us more flexibility. That also created some extra features and challenges to be incorporated in Debezium itself!
```